tutorials/language/tensorize.py

```diff
@@ -0,0 +1,276 @@
+# We compare the tensorize version with that :code:`numpy.dot` produces,
+# ensure our implementation is correct.
+#
+# Reduce-update for Tensorize
+# ------------------------------------
+# Let's then move one step forward.
+# Assume our accelerator could only multiply a vector by a square matrix,
+# in which the vector size needs to be no larger than 16.
+# Given such hardware constrain, now we need to split the reduce axis as following,
+#
+zo, zi = s[C].split(z, factor=factor)
+s[C].reorder(x, yo, zo, yi, zi)
+
+######################################################################
+# However, since the tensorize intrinsic now only covers a part of the reduce axis,
+# instead of using one "body" function, TVM requires a :code:`reduce_reset` function,
+# which will be invoked before the reduce for-loop, and a :code:`reduce_update` function,
+# which defines the "update" computing strategy.
+#
+def gemv_impl():
+    cc_code = """
+      extern "C" int gemv_update(float *cc, float *aa, float *bb, int m, int l, int stride) {
+        for (int i = 0; i < m; ++i) {
+            for (int j = 0; j < l; ++j) {
+                cc[i] += aa[j] * bb[i * stride + j];
+            }
+        }
+        return 0;
+      }
+      extern "C" int gemv_reset(float *cc, int m) {
+        for (int i = 0; i < m; ++i) {
+            cc[i] = 0.0;
+        }
+        return 0;
+      }
+    """
+    from tvm.contrib import util, clang
+    temp = util.tempdir()
+    ll_path = temp.relpath("temp.ll")
+    # Create LLVM ir from c source code
+    ll_code = clang.create_llvm(cc_code, output=ll_path)
+    return ll_code
+
+def intrin_gemv(m, l):
+    a = tvm.placeholder((l,), name='a')
+    b = tvm.placeholder((m, l), name='b')
+    k = tvm.reduce_axis((0, l), name='k')
+    c = tvm.compute((m,), lambda i:
+    tvm.sum(a[k] * b[i, k], axis=k), name='c')
+    Ab = tvm.decl_buffer(a.shape, a.dtype,
+                         name="A",
+                         offset_factor=1,
+                         strides=[1])
+    Bb = tvm.decl_buffer(b.shape, b.dtype,
+                         name="B",
+                         offset_factor=1,
+                         strides=[tvm.var("ldw"), 1])
+    Cb = tvm.decl_buffer(c.shape, c.dtype,
+                         name="C",
+                         offset_factor=1,
+                         strides=[1])
+    def intrin_func(ins, outs):
+        def _instrin(index):
+            ib = tvm.ir_builder.create()
+            aa, bb = ins
+            cc = outs[0]
+            if index == 1:
+                ib.emit(tvm.call_extern("int32", "gemv_reset", cc.access_ptr("w"), m))
+            else:
+                ib.emit(tvm.call_extern("int32", "gemv_update",
+                                        cc.access_ptr("w"), aa.access_ptr("r"), bb.access_ptr("r"),
+                                        m, l, bb.strides[0]))
+            return ib.get()
+        return _instrin(0), _instrin(1), _instrin(2)
```

Comment:
Here, I think we need more details here. We can rewrite the code to have 3 functions - reduce_reset, reduce_update and body. This will make things clearer. I think this is important for first-time users. I had the same questions when I was trying to use tensorize.